Добавила новый пакет request

npoval opened the pull request:
Добавила 2 новых класса LoanRequest и LoanType
Убрала использование static в классах Repository, Service, Controller

Commit:
Добавила новый пакет request
Добавила 2 новых класса LoanRequest и LoanType
Убрала использование static в классах Repository, Service, Controller

neparij commented on app/src/main/java/com/tinkoff/edu/app/logger/LoanCalcLogger.java:
```diff
@@ -1,11 +1,14 @@
      * TODO Logger
      */
-    public static void log() {
+    public static void log(LoanRequest request) {
```
request не используется.
В логгере можно использовать данные из request и отобразить их. Тем более в LoanRequest реализован метод toString()

npoval replied:
поправила, но в проекте Евгений пока что удалил логгер совсем

Commit:
Add:
non-static Controller with Service DI
non-static Service with Repo DI
Service with Interface
non-static Repository with Interface
non-static Loan Request
non-static Loan Response with encapsulated autogenerated incremental id
Loan Response encapsulates Response Type enum: APPROVED, DENIED

Commit:
Написано 2 теста:
shouldGet1WnenNotIpRequest
shouldGetIncrementedIdWnenAnyCall

Commit:
Доработан тест с циклом:
shouldGetIncrementedIdWnenAnyCall
Создан новый тест LoanResponseTest
Исправлена реализация класса LoanResponse

Commit:
Исправила название переменной sut на controller

Commit:
Исправила название метода

Commit:
Добавлены бизнес-правила для контроллера
Доработан класс LoanResponse
Убран статик из репо
Написаны тесты, покрывающие бизнес-правила
Покрытие по веткам контроллера 92%

neparij commented on app/src/main/java/com/tinkoff/edu/app/controller/LoanCalcController.java:
```diff
@@ -1,19 +1,35 @@
-        LoanCalcLogger.log();
-        return LoanCalcService.createRequest();
+        //LoanCalcLogger.log(request);
+        if ((request != null && request.getType().equals(LoanType.PERSON) && (request.getAmount() > 0 && request.getAmount() <= 10000) && ((request.getMonths() > 0) && request.getMonths() <= 12)) || ((request != null && request.getType().equals(LoanType.OOO)) && (request.getAmount() > 10000) && (request.getMonths() > 0 && request.getMonths() < 12))) {
```
Ну очень длинная строка получилась. Надо упрощать и пытаться найти общие условия.

npoval replied:
исправила сразу все на switch - все равно нужно для новой ДЗ

neparij commented on app/src/main/java/com/tinkoff/edu/app/controller/LoanCalcController.java:
```diff
@@ -1,19 +1,35 @@
+            return new LoanResponse(ResponseType.APPROVED, service.createRequest(request));
+        } else if (request == null || request.getAmount() <= 0 || request.getMonths() <= 0) {
+            return new LoanResponse(ResponseType.ERROR, -1);
+        } else
```
по соглашению оформления кода мы не можем писать код выполняемый условным оператором без фигурных скобок.
Правильно будет:
else {
    // ... исполняемый блок кода
}

npoval replied:
исправила в новой версии со switch

Commit:
Переделала реализацию бизнес логики на более читаемую и понятную и сразу с использованием switch для ДЗ

neparij commented on app/src/main/java/com/tinkoff/edu/app/response/LoanResponse.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+
+    public ResponseType getResponse() {
```
Название геттера не соответствует названию поля. getResponseType будет актуальным

npoval replied:
исправила

neparij commented on app/src/main/java/com/tinkoff/edu/app/response/LoanResponse.java:
```diff
@@ -0,0 +1,40 @@
+ */
+
+public class LoanResponse {
+    //private static int count = 0;
```
Если код не нужен - его можно смело удалять.

npoval replied:
исправила

neparij commented on app/src/main/java/com/tinkoff/edu/app/response/LoanResponse.java:
```diff
@@ -0,0 +1,40 @@
+public class LoanResponse {
+    //private static int count = 0;
+    private ResponseType responseType;
+    private  int responseId;
```
```suggestion
    private int responseId;
```

npoval replied:
исправлено

neparij commented on app/src/main/java/com/tinkoff/edu/app/response/LoanResponse.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(responseType, getResponseId());
```
```suggestion
        return Objects.hash(responseType, responseId);
```
либо
```suggestion
        return Objects.hash(getResponseType(), getResponseId());
```

Надо определиться мы будем использовать методы или нет

npoval replied:
исправила, методы не юзаются больше

neparij reviewed (commented):
В LoanCalcTest объект request созданный в init() используется в 2 из 16 тестов.

neparij commented on app/pom.xml:
```diff
@@ -23,51 +23,46 @@
       <version>5.8.0-M1</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
зачем зависимость junit:junit?
Мы пятый JUnit подключаем через org.junit.jupiter, здесь же добавится четвертый и не будет использоваться в проекте, я не нашел вызовов методов из этого пакета

npoval replied:
исправила

Commit:
Добавлена генерация идентификатора через UUID
Добавлен новый класс ArrayRepo
Исправлен тип идентификатора с int на UUID
Убран неиспользуемый код
Тесты разнесены по классам
Тесты отрефакторены

Commit:
Доработан контроллер
Добавлено использование эксепшенов
Создано 3 своих экспешнена
Добавлена проверка на длину fio
Произведен рефакторинг тестов
Написаны тесты на эксепшены
Добавлены параметризованные тесты

neparij commented on app/src/main/java/com/tinkoff/edu/app/controller/LoanCalcController.java:
```diff
@@ -1,19 +1,54 @@
-}
 
+    public LoanResponse createRequest(LoanRequest request) throws FioDataLengthException, MonthsDataValidationException, AmountValidationException {
+        if (request != null && request.getFio() != null) {
```
request.getType() тоже может оказать null

neparij commented on app/src/main/java/com/tinkoff/edu/app/exception/FioDataLengthException.java:
```diff
@@ -0,0 +1,7 @@
+package com.tinkoff.edu.app.exception;
+
+public class FioDataLengthException extends Exception {
```
Можно также переименовать как и другие исключения валидации FioValidationException для однообразности

neparij commented on app/src/main/java/com/tinkoff/edu/app/controller/LoanCalcController.java:
```diff
@@ -1,19 +1,54 @@
+                    throw new IllegalArgumentException("Некорректные данные!");
+            }
+        } else {
+            throw new NullPointerException("Пустые данные в запросе!");
```
Возможно здесь стоит сделать также IllegalArgumentException. А то получается ситуация, что мы проверяем на null, чтобы не выкинуть NullPointerException, а потом всё равно его кидаем :)

neparij commented on app/src/main/java/com/tinkoff/edu/app/controller/LoanCalcController.java:
```diff
@@ -1,19 +1,54 @@
     }
-}
 
+    public LoanResponse createRequest(LoanRequest request) throws FioDataLengthException, MonthsDataValidationException, AmountValidationException {
```
Исключения валидации можно сделать одним родительским исключением и наследоваться от него.
Т.е. создать класс ValidationException (от Exception) и от него наследовать уже остальные
например: `public class AmountValidationException extends ValidationException {`
Ну и отлавливать после можно будет лишь ValidationException, даже в случае, когда у тебя появится много других валидаций

neparij commented on app/src/main/java/com/tinkoff/edu/app/repository/ArrayLoanCalcRepository.java:
```diff
@@ -0,0 +1,47 @@
+    static int counter;
+
+    public ResponseType getStatus(UUID responseId) {
+        for (int i = 0; i < responses.length; i++) {
```
Можно использовать enhanced for, если интересно почитай тут: https://blogs.oracle.com/corejavatechtips/using-enhanced-for-loops-with-your-classes

neparij commented on app/src/main/java/com/tinkoff/edu/app/repository/ArrayLoanCalcRepository.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public void setStatus(UUID responseId, ResponseType responseType) {
+        for (int i = 0; i < responses.length; i++) {
```
Можно использовать enhanced for, если интересно почитай тут: https://blogs.oracle.com/corejavatechtips/using-enhanced-for-loops-with-your-classes

Commit:
Добавлен MapRepo
Произведен рефакторинг тестов с новым репо
СОздан новый класс LoanRequestResponse, для хранения в Map информации по заявке
Написаны тесты на новый репо

Commit:
Добавлен FileRepo
Произведен рефакторинг тестов для контроллера с новым file repo
Написаны тесты на новый file repo
Изменены названия классов для тестов контроллера

neparij commented on app/src/test/java/com/tinkoff/edu/ArrayRepoTest.java:
```diff
@@ -0,0 +1,67 @@
+        LoanResponse response = controller.createRequest(this.request);
+        repository.writeResponse(response);
+        System.out.println(response.getResponseId());
+        repository.getStatus(response.getResponseId());
```
строка не имеет смысла, так как полученное значение не используется

neparij commented on app/src/test/java/com/tinkoff/edu/ArrayRepoTest.java:
```diff
@@ -0,0 +1,67 @@
+    @DisplayName("Проверка возврата NOTFOUND при получении статуса заявки по несуществующему UUID")
+    public void shouldGetLoanStatusByNotFoundUuid() {
+        UUID uuid = UUID.fromString("123e4567-e89b-12d3-a456-4266554400");
+        repository.getStatus(uuid);
```
строка не имеет смысла, так как полученное значение не используется

neparij commented on app/src/test/java/com/tinkoff/edu/ControllerTestWithArrayRepo.java:
```diff
@@ -0,0 +1,151 @@
+
+    @Test
+    @DisplayName("Проверка возврата ошибки Некорректное значение месяца! при отрицательном значении месяца")
+    public void shouldGetErrorWhenNegativeMonthsRequest() {
```
Не критично, но в целом все тесты на исключения валидации можно также сделать параметризованными

neparij commented on app/src/test/java/com/tinkoff/edu/ControllerTestWithFileRepo.java:
```diff
@@ -0,0 +1,138 @@
+    @ParameterizedTest
+    @MethodSource("provideInValidMonthsRequestAndResponse")
+    @DisplayName("Проверка возврата ошибки Некорректное значение месяца! при некорретных значениях месяца")
+    public void shouldGetErrorWhenNegativeMonthsRequest2(LoanType LoanType, int months, int amount, String fio, String expectedMessage) {
```
Все тесты валидации можно сделать параметризованными от одного источника данных.
Можно передавать как expectedMessage, так и класс исключения

neparij commented on app/src/test/java/com/tinkoff/edu/ControllerTestWithFileRepo.java:
```diff
@@ -0,0 +1,138 @@
+        controller = new LoanCalcController(new LoanCalcService(repository));
+        final FioDataLengthException thrown = assertThrows(FioDataLengthException.class,
+                () -> controller.createRequest(request));
+        assertEquals("Некорректная длина FIO!", thrown.getMessage());
```
expectedMessage?

neparij commented on app/src/test/java/com/tinkoff/edu/ControllerTestWithMapRepo.java:
```diff
@@ -0,0 +1,135 @@
+        controller = new LoanCalcController(new LoanCalcService(repository));
+        final FioDataLengthException thrown = assertThrows(FioDataLengthException.class,
+                () -> controller.createRequest(request));
+        assertEquals("Некорректная длина FIO!", thrown.getMessage());
```
expectedMessage?

neparij commented on app/src/test/java/com/tinkoff/edu/ControllerTestWithMapRepo.java:
```diff
@@ -0,0 +1,135 @@
+    @ParameterizedTest
+    @MethodSource("provideInValidMonthsRequestAndResponse")
+    @DisplayName("Проверка возврата ошибки Некорректное значение месяца! при некорретных значениях месяца")
+    public void shouldGetErrorWhenNegativeMonthsRequest2(LoanType LoanType, int months, int amount, String fio, String expectedMessage) {
```
Все тесты валидации можно сделать параметризованными от одного источника данных.
Можно передавать как expectedMessage, так и класс исключения

neparij commented on app/src/test/java/com/tinkoff/edu/ControllerTestWithMapRepo.java:
```diff
@@ -0,0 +1,135 @@
+    @ParameterizedTest
+    @MethodSource("provideInValidMonthsRequestAndResponse")
+    @DisplayName("Проверка возврата ошибки Некорректное значение месяца! при некорретных значениях месяца")
+    public void shouldGetErrorWhenNegativeMonthsRequest2(LoanType LoanType, int months, int amount, String fio, String expectedMessage) {
```
Зачем "2" в названии метода?

neparij replied:
аналогично в других местах

neparij commented on app/src/test/java/com/tinkoff/edu/MapRepoTest.java:
```diff
@@ -0,0 +1,93 @@
+        for (int i = 0; i < 3; i++) {
+            LoanResponse response = controller.createRequest(this.request);
+            loanRequestResponse = new LoanRequestResponse(response, request);
+            repository.setLoanRequestResponseMap(response.getResponseId(),loanRequestResponse);}
```
"}" должно быть на следующей строке.

neparij commented on app/src/test/java/com/tinkoff/edu/MapRepoTest.java:
```diff
@@ -0,0 +1,93 @@
+        for (int i = 0; i < 3; i++) {
+            LoanResponse response = controller.createRequest(this.request);
+            loanRequestResponse = new LoanRequestResponse(response, request);
+            repository.setLoanRequestResponseMap(response.getResponseId(),loanRequestResponse);}
```
"}" должно быть на следующей строке.

neparij commented on app/src/test/java/com/tinkoff/edu/MapRepoTest.java:
```diff
@@ -0,0 +1,93 @@
+        request = new LoanRequest(LoanType.PERSON, 10, 1000, "Русаков Иван Иванович-Петрович");
+        repository = new MapLoanCalcRepository();
+        controller = new LoanCalcController(new LoanCalcService(repository));
+
```
лишний перенос

neparij commented on app/src/main/java/com/tinkoff/edu/app/request/LoanRequest.java:
```diff
@@ -0,0 +1,45 @@
+package com.tinkoff.edu.app.request;
+
+/**
+ * Class, Type-> objects,instances
```
Незаполненный javadoc

neparij commented on app/src/main/java/com/tinkoff/edu/app/repository/FileLoanCalcRepo.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public String getStatusByUuid(UUID uuid) throws IOException {
+        final String[] status = new String[1];
```
Зачем здесь массив строк?
Не  понял немного решение.
